### PR TITLE
Add parameter interventions to calibrate and refactor interface tests

### PR DESCRIPTION
**Pull request by SamWitty**
This PR adds parameter intervention to the `calibrate` interface, and refactors the interface tests to be more modular.

Here, adding an intervention to `calibrate` refers to the scenario where we are conditioning on data that was drawn from an interventional distribution. In other words, this is the continuous-time analog of how experimental data is incorporated into probabilistic inferences in this paper: https://arxiv.org/abs/1910.14124

### CODE REVIEW

**Review by djinnome — COMMENTED** [review]
Quick question: how is this different from intervening on a model, and then sending the intervened model to calibrate, along with post-intervention data?

**Comment by SamWitty** [comment]
> Quick question: how is this different from intervening on a model, and then sending the intervened model to calibrate, along with post-intervention data?

This is essentially what is happening. We could further separate out `intervene` from `calibrate` to accomplish this more directly, but currently the interface methods only operate on AMRs directly, not the compiled model.

File: pyciemss/interfaces.py [hunk 1]
```diff
@@ -273,17 +273,24 @@ def wrapped_model():
...
-        # Adding deterministic nodes to the model so that we can access the trajectory in the Predictive object.
-        [pyro.deterministic(k, v) for k, v in trajectory.items()]
+        # Need to add observables to the trajectory, as well as add deterministic nodes to the model.
+        trajectory_observables = model.observables(trajectory)
+        [
+            pyro.deterministic(f"{k}_observable", v)
+            for k, v in trajectory_observables.items()
+        ]
+
+        full_trajectory = {**trajectory, **trajectory_observables}
```
[the remark this answers is not in the document]
Comment (SamWitty) [reply pyciemss/interfaces.py:286]: This is a bit subtle. The `LogTrajectory` handler only handles the `simulate` operation, and not any computation downstream of `simulate`. As `simulate` only provides the state variables, and not the observables, we need to apply the observables function again on the induced trajectory (which only contains states) to add them in.

File: pyciemss/interfaces.py [hunk 2]
```diff
@@ -302,16 +309,22 @@ def calibrate(
     solver_method: str = "dopri5",
     solver_options: Dict[str, Any] = {},
     start_time: float = 0.0,
-    static_interventions: Dict[float, Dict[str, torch.Tensor]] = {},
-    dynamic_interventions: Dict[
-        Callable[[Dict[str, torch.Tensor]], torch.Tensor], Dict[str, torch.Tensor]
+    static_state_interventions: Dict[torch.Tensor, Dict[str, Intervention]] = {},
+    static_parameter_interventions: Dict[torch.Tensor, Dict[str, Intervention]] = {},
+    dynamic_state_interventions: Dict[
```
[the remark this answers is not in the document]
Comment (SamWitty) [reply pyciemss/interfaces.py:314]: That's correct. When calling `calibrate` a user passes in interventions that describe the process by which the data was generated. How they use the inferred parameters after the fact is totally up to them. They could `sample` using different interventions, they could `sample` with the same interventions into the future, etc.

File: tests/test_interfaces.py [hunk 3]
```diff
@@ -197,79 +196,147 @@ def intervention_event_fn_2(time: torch.Tensor, *args, **kwargs):
...
     with pyro.poutine.seed(rng_seed=0):
-        calibrated_result_2 = sample(
-            model_url,
-            end_time,
-            logging_step_size,
-            1,
-            start_time=start_time,
-            inferred_parameters=inferred_parameters_2,
-        )["unprocessed_result"]
+        intervened_parameters, intervened_loss = calibrate(
```
[the remark this answers is not in the document]
Comment (SamWitty) [reply tests/test_interfaces.py:329]: There is only one set of learned parameters. The interventions just change the dynamics, which implicitly changes the likelihood of observed data.

**Comment by djinnome** [comment]
Gotcha!  So, previously, we could only sample from a conditional distribution (by sampling from an observational model with inferred parameters from an observational distribution), from a counterfactual distribution (by sampling from an intervened model with inferred parameters from an observational distribution), but we couldn't sample from a conditional post-intervention distribution (sample from intervened model using  inferred parameters from an interventional distribution).

It seems you could also sample from an observational model using inferred parameters from an interventional distribution, or sample from an intervened model using inferred parameters from a different interventional distribution, too!

**Review by djinnome — APPROVED** [review]
Thanks for the  clarifications, @SamWitty. Looks good to me!

**Pull request merged**